File: src/agent/tools/builtin/bash.py
"""Bash tool: execute shell commands."""

from __future__ import annotations

import asyncio
from typing import Any


class BashTool:
    def __init__(self, timeout: int = 120) -> None:
        self._timeout = timeout

    @property
    def name(self) -> str:
        return "bash"

    @property
    def description(self) -> str:
        return "Execute a shell command and return its output."

    @property
    def input_schema(self) -> dict[str, Any]:
        return {
            "type": "object",
            "properties": {
                "command": {
                    "type": "string",
                    "description": "The shell command to execute.",
                },
                "timeout": {
                    "type": "integer",
                    "description": f"Timeout in seconds (default: {self._timeout}).",
                },
            },
            "required": ["command"],
        }

    async def execute(self, *, command: str, timeout: int | None = None, **_: Any) -> str:
        effective_timeout = timeout if timeout is not None else self._timeout
        try:
            proc = await asyncio.create_subprocess_shell(
                command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await asyncio.wait_for(
                proc.communicate(), timeout=effective_timeout,
            )
        except asyncio.TimeoutError:
            proc.kill()
            await proc.wait()
            return f"Command timed out after {effective_timeout}s"

        output_parts = []
        if stdout:
            output_parts.append(stdout.decode(errors="replace"))
        if stderr:
            output_parts.append(f"STDERR:\n{stderr.decode(errors='replace')}")
        if proc.returncode != 0:
            output_parts.append(f"Exit code: {proc.returncode}")

        result = "\n".join(output_parts)
        return result or "(no output)"

File: src/tools/builtin/bash.ts
/** Bash tool: execute shell commands. */

import { spawn } from 'node:child_process';

import type { Tool } from '../base.js';

export class BashTool implements Tool {
  constructor(private timeout = 120) {}

  get name(): string {
    return 'bash';
  }

  get description(): string {
    return 'Execute a shell command and return its output.';
  }

  get inputSchema(): Record<string, any> {
    return {
      type: 'object',
      properties: {
        command: {
          type: 'string',
          description: 'The shell command to execute.',
        },
        timeout: {
          type: 'integer',
          description: `Timeout in seconds (default: ${this.timeout}).`,
        },
      },
      required: ['command'],
    };
  }

  async execute(params: Record<string, any>): Promise<string> {
    const command: string = params.command;
    const effectiveTimeout: number =
      typeof params.timeout === 'number' ? params.timeout : this.timeout;

    return new Promise<string>((resolve) => {
      const proc = spawn(command, { shell: true });
      const stdoutChunks: Buffer[] = [];
      const stderrChunks: Buffer[] = [];
      let timedOut = false;

      const timer = setTimeout(() => {
        timedOut = true;
        proc.kill('SIGKILL');
      }, effectiveTimeout * 1000);

      proc.stdout.on('data', (d: Buffer) => stdoutChunks.push(d));
      proc.stderr.on('data', (d: Buffer) => stderrChunks.push(d));

      proc.on('error', (err) => {
        clearTimeout(timer);
        resolve(`Error executing command: ${err.message}`);
      });

      proc.on('close', (code) => {
        clearTimeout(timer);
        if (timedOut) {
          resolve(`Command timed out after ${effectiveTimeout}s`);
          return;
        }
        const stdout = Buffer.concat(stdoutChunks).toString('utf-8');
        const stderr = Buffer.concat(stderrChunks).toString('utf-8');
        const parts: string[] = [];
        if (stdout) parts.push(stdout);
        if (stderr) parts.push(`STDERR:\n${stderr}`);
        if (code !== 0 && code !== null) parts.push(`Exit code: ${code}`);
        resolve(parts.join('\n') || '(no output)');
      });
    });
  }
}
